### ISSUE TITLE
slicSuperpixels() always returns '0'.

### ISSUE THREAD

**Issue by RishabhMalviya** [ukoethe/vigra#281]
I used the library (1.10.0 release) as part of a [project](https://github.com/RishabhMalviya/Semantic_Perception) recently, and I think I may have discovered a bug with the `slicSuperpixels()` function. It states the following in the [documentation](http://ukoethe.github.io/vigra/doc-release/vigra/group__SeededRegionGrowing.html#ga5561f1ce9b3f7a95b7de67738a0d2447) on the website: 

*"The function returns the number of superpixels, which equals the largest label because labeling starts at 1."*

But when using the function, it always returned '0'. I found a way around this by calculating the maximum value in the `labels` MultiArray that is filled up by the function (please look at how the variable `superpixelCount` is assigned in the `extractSLICSuperpixels` function at line number 59 in [this](https://github.com/RishabhMalviya/Semantic_Perception/blob/master/src/features_slic.cpp) source file). Furthermore, when going through these super-pixels, I discovered that many of them had zero pixels in them. 

I think this may be because internally some superpixels get merged with others. I do not know exactly what files to work on or how to proceed to fix his issue, but I felt that I should bring this to light so that it can be worked upon.

**Comment by ukoethe** [comment]
Thanks for reporting this. Can you please try if the problem persists when you use the current master from github? 

> Furthermore, when going through these super-pixels, I discovered that many of them had zero pixels in them. I think this may be because internally some superpixels get merged with others.

According to the original authors of the SLIC method, merging of some superpixels is desired. However, function `slicSuperpixels()` is supposed to call function `Slic::postProcessing()` which should restore consecutive superpixel labels and return the maximum. Is this function called in your installation?

**Comment by RishabhMalviya** [comment]
Actually, I haven't tried this with the latest installation.

Is it fine if I get back to you with my results within two days? I am a
little busy at the moment.

On Tue, Dec 22, 2015 at 2:10 AM, Ullrich Koethe <notifications@github.com>
wrote:

> Thanks for reporting this. Can you please try if the problem persists when
> you use the current master from github?
>
> Furthermore, when going through these super-pixels, I discovered that many
> of them had zero pixels in them. I think this may be because internally
> some superpixels get merged with others.
>
> According to the original authors of the SLIC method, merging of some
> superpixels is desired. However, function slicSuperpixels() is supposed
> to call function Slic::postProcessing() which should restore consecutive
> superpixel labels and return the maximum. Is this function called in your
> installation?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ukoethe/vigra/issues/281#issuecomment-166411318>.
>



-- 
Rishabh Malviya
Manager
Innovation Cell
IIT Bombay

**Comment by RishabhMalviya** [comment]
I just ran some test with the latest snapshot from GitHub. **There aren't any problems with the slicSuperpixels() function any more!**

I ran the function on the *lenna_color.gif* image, and I got 61 superpixels, which the function returned faithfully. I also verified that none of the superpixels had zero pixels in them by calculating the average RGB values for each through an ArrayOfRegionStatistics object (when I did this previously, I was getting 'nan' values for the zero-pixel superpixels). I've attached a screenshot of my results:

![screenshot from 2015-12-22 13 14 36](https://cloud.githubusercontent.com/assets/6564529/11946428/a8d3493e-a880-11e5-86a8-19688173cd04.png)

I also checked the *slic.hxx* file and verified that the `Slic::postProcessing()` function is being called.
